### ISSUE TITLE
Create DocumentValidator and move validations targeting document header

### ISSUE THREAD

**Issue by jonasdmoeller** [Energinet-DataHub/geh-charges#1305]
**Description**
Validation of document header (which becomes the DocumentDto object of ChargeCommand) needs to be separated from input and business validation, since a failed validation in the header should result in rejection of all operations in the message.
This could potentially be done only once for the entire request, but that is not in scope for this task.

**AC**

- [ ] Validation errors of following rules must send rejections for all operations in message:

* CommandSenderMustBeAnExistingMarketParticipantRule
* BusinessReasonCodeMustBeUpdateChargeInformationRule
* DocumentTypeMustBeRequestChangeOfPriceListRule
* RecipientIsMandatoryTypeValidationRule 
* SenderIsMandatoryTypeValidationRule

![image.png](https://images.zenhubusercontent.com/60dafff3568c156496e54f35/8933a4b6-9c82-469d-a27f-29f28b87e053)

**Comment by prtandrup** [comment]
### Tested in T-001.

:warning: Not all valdations can be triggered due to schema validation and exceptions in the xml deserializer if GLN numbers are empty.

**AC 1:**

- [ ] **CommandSenderMustBeAnExistingMarketParticipantRule**: (N/A - blocked by sync validation, B2B-008)

- [x] **BusinessReasonCodeMustBeUpdateChargeInformationRule**:

Request sent, which uses incorrect Business Reason Code = 'D17'

````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
	<cim:mRID>DocId2022-05-10T07:45:16.116Z</cim:mRID>
	<cim:type>D10</cim:type>
	<cim:process.processType>D17</cim:process.processType>
	<cim:businessSector.type>23</cim:businessSector.type>
	<cim:sender_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:sender_MarketParticipant.mRID>
	<cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
	<cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
	<cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
	<cim:createdDateTime>2022-05-10T09:45:15Z</cim:createdDateTime>
	<cim:MktActivityRecord>
		<cim:mRID>OpId2022-05-10T07:45:16.116Z</cim:mRID>
		<cim:ChargeGroup>
			<cim:ChargeType>
				<cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:chargeTypeOwner_MarketParticipant.mRID>
				<cim:type>D02</cim:type>
				<cim:mRID>FeeGrid001</cim:mRID>
				<cim:name>Charge Fee Name</cim:name>
				<cim:description>The charge description</cim:description>
				<cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
				<cim:effectiveDate>2022-07-28T22:00:00Z</cim:effectiveDate>
				<cim:VATPayer>D01</cim:VATPayer>
				<cim:transparentInvoicing>false</cim:transparentInvoicing>
				<cim:taxIndicator>false</cim:taxIndicator>
			</cim:ChargeType>
		</cim:ChargeGroup>
	</cim:MktActivityRecord>
    	<cim:MktActivityRecord>
		<cim:mRID>OpId22022-05-10T07:45:16.116Z</cim:mRID>
		<cim:ChargeGroup>
			<cim:ChargeType>
				<cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:chargeTypeOwner_MarketParticipant.mRID>
				<cim:type>D02</cim:type>
				<cim:mRID>FeeGrid002</cim:mRID>
				<cim:name>Charge Fee Name</cim:name>
				<cim:description>The charge description</cim:description>
				<cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
				<cim:effectiveDate>2022-07-28T22:00:00Z</cim:effectiveDate>
				<cim:VATPayer>D01</cim:VATPayer>
				<cim:transparentInvoicing>false</cim:transparentInvoicing>
				<cim:taxIndicator>false</cim:taxIndicator>
			</cim:ChargeType>
		</cim:ChargeGroup>
	</cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

Peeked response, containing both Operations:

````
<?xml version="1.0" encoding="utf-8"?>
<cim:RejectRequestChangeOfPriceList_MarketDocument xmlns:cim="urn:ediel.org:structure:rejectrequestchangeofpricelist:0:1">
    <cim:mRID>b8b5f1e879b5495cbc6d8470297b6b63</cim:mRID>
    <cim:type>D11</cim:type>
    <cim:process.processType>D17</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDZ</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDM</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>2022-05-10T07:46:06Z</cim:createdDateTime>
    <cim:reason.code>A02</cim:reason.code>
    <cim:MktActivityRecord>
        <cim:mRID>99715563dc874279ab3eeed9590c1d62</cim:mRID>
        <cim:originalTransactionIDReference_MktActivityRecord.mRID>OpId22022-05-10T07:45:16.116Z</cim:originalTransactionIDReference_MktActivityRecord.mRID>
        <cim:Reason>
            <cim:code>D02</cim:code>
            <cim:text>Energy business process UpdateMasterDataSettlement not allowed together with document type RequestChangeOfPriceList</cim:text>
        </cim:Reason>
    </cim:MktActivityRecord>
    <cim:MktActivityRecord>
        <cim:mRID>1df7c9b598264d4eb57086b08d3cff3d</cim:mRID>
        <cim:originalTransactionIDReference_MktActivityRecord.mRID>OpId2022-05-10T07:45:16.116Z</cim:originalTransactionIDReference_MktActivityRecord.mRID>
        <cim:Reason>
            <cim:code>D02</cim:code>
            <cim:text>Energy business process UpdateMasterDataSettlement not allowed together with document type RequestChangeOfPriceList</cim:text>
        </cim:Reason>
    </cim:MktActivityRecord>
</cim:RejectRequestChangeOfPriceList_MarketDocument>
````

- [x] **DocumentTypeMustBeRequestChangeOfPriceListRule**:

Request sent, with wrong inbound document type code D12 (Which is notify price list):

````
<?xml version="1.0" encoding="UTF-8"?>
<cim:RequestChangeOfPriceList_MarketDocument xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:cim="urn:ediel.org:structure:requestchangeofpricelist:0:1" xsi:schemaLocation="urn:ediel.org:structure:requestchangeofpricelist:0:1 urn-ediel-org-structure-requestchangeofpricelist-0-1.xsd">
	<cim:mRID>DocId2022-05-10T07:48:43.487Z</cim:mRID>
	<cim:type>D12</cim:type>
	<cim:process.processType>D18</cim:process.processType>
	<cim:businessSector.type>23</cim:businessSector.type>
	<cim:sender_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:sender_MarketParticipant.mRID>
	<cim:sender_MarketParticipant.marketRole.type>DDM</cim:sender_MarketParticipant.marketRole.type>
	<cim:receiver_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:receiver_MarketParticipant.mRID>
	<cim:receiver_MarketParticipant.marketRole.type>DDZ</cim:receiver_MarketParticipant.marketRole.type>
	<cim:createdDateTime>2022-05-10T09:48:43Z</cim:createdDateTime>
	<cim:MktActivityRecord>
		<cim:mRID>OpId2022-05-10T07:48:43.487Z</cim:mRID>
		<cim:ChargeGroup>
			<cim:ChargeType>
				<cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:chargeTypeOwner_MarketParticipant.mRID>
				<cim:type>D02</cim:type>
				<cim:mRID>FeeGrid001</cim:mRID>
				<cim:name>Charge Fee Name</cim:name>
				<cim:description>The charge description</cim:description>
				<cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
				<cim:effectiveDate>2022-07-28T22:00:00Z</cim:effectiveDate>
				<cim:VATPayer>D01</cim:VATPayer>
				<cim:transparentInvoicing>false</cim:transparentInvoicing>
				<cim:taxIndicator>false</cim:taxIndicator>
			</cim:ChargeType>
		</cim:ChargeGroup>
	</cim:MktActivityRecord>
    	<cim:MktActivityRecord>
		<cim:mRID>OpId22022-05-10T07:48:43.487Z</cim:mRID>
		<cim:ChargeGroup>
			<cim:ChargeType>
				<cim:chargeTypeOwner_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:chargeTypeOwner_MarketParticipant.mRID>
				<cim:type>D02</cim:type>
				<cim:mRID>FeeGrid002</cim:mRID>
				<cim:name>Charge Fee Name</cim:name>
				<cim:description>The charge description</cim:description>
				<cim:priceTimeFrame_Period.resolution>P1D</cim:priceTimeFrame_Period.resolution>
				<cim:effectiveDate>2022-07-28T22:00:00Z</cim:effectiveDate>
				<cim:VATPayer>D01</cim:VATPayer>
				<cim:transparentInvoicing>false</cim:transparentInvoicing>
				<cim:taxIndicator>false</cim:taxIndicator>
			</cim:ChargeType>
		</cim:ChargeGroup>
	</cim:MktActivityRecord>
</cim:RequestChangeOfPriceList_MarketDocument>
````

Peeked response, rejecting both operations:

````
<?xml version="1.0" encoding="utf-8"?>
<cim:RejectRequestChangeOfPriceList_MarketDocument xmlns:cim="urn:ediel.org:structure:rejectrequestchangeofpricelist:0:1">
    <cim:mRID>2c2d13e1dac7488bb21637d6072dfc14</cim:mRID>
    <cim:type>D11</cim:type>
    <cim:process.processType>D18</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDZ</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDM</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>2022-05-10T07:49:20Z</cim:createdDateTime>
    <cim:reason.code>A02</cim:reason.code>
    <cim:MktActivityRecord>
        <cim:mRID>5e18b0f967424d28a56b5a5c06c615d5</cim:mRID>
        <cim:originalTransactionIDReference_MktActivityRecord.mRID>OpId22022-05-10T07:48:43.487Z</cim:originalTransactionIDReference_MktActivityRecord.mRID>
        <cim:Reason>
            <cim:code>D02</cim:code>
            <cim:text>Document type NotifyPriceList not allowed together with energy business process UpdateChargeInformation</cim:text>
        </cim:Reason>
    </cim:MktActivityRecord>
    <cim:MktActivityRecord>
        <cim:mRID>2b6ff9e8fb2642e680e4a4a94bd249fa</cim:mRID>
        <cim:originalTransactionIDReference_MktActivityRecord.mRID>OpId2022-05-10T07:48:43.487Z</cim:originalTransactionIDReference_MktActivityRecord.mRID>
        <cim:Reason>
            <cim:code>D02</cim:code>
            <cim:text>Document type NotifyPriceList not allowed together with energy business process UpdateChargeInformation</cim:text>
        </cim:Reason>
    </cim:MktActivityRecord>
</cim:RejectRequestChangeOfPriceList_MarketDocument>
````

- [ ] **RecipientIsMandatoryTypeValidationRule**: N/A - 500 internal error: deserializer cannot read an empty value

- [ ] **SenderIsMandatoryTypeValidationRule**: N/A - 500 internal error: deserializer cannot read an empty value

**Comment by prtandrup** [comment]
Need to fix the error texts using the words "energy business process". Is it the correct wording to use for business reason code for DH3?

**Comment by prtandrup** [comment]
Need to fix the error texts using the words "energy business process". Is it the correct wording to use for business reason code for DH3?

**Comment by prtandrup** [comment]
### Repeated a few test in T-001 after changing error texts
Tested :white_check_mark: - moving to closed.

*Used business reason code D17, response:*

````
<?xml version="1.0" encoding="utf-8"?>
<cim:RejectRequestChangeOfPriceList_MarketDocument xmlns:cim="urn:ediel.org:structure:rejectrequestchangeofpricelist:0:1">
    <cim:mRID>79aa2c49210141018baa353d79691b9f</cim:mRID>
    <cim:type>D11</cim:type>
    <cim:process.processType>D17</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDZ</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDM</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>2022-05-16T10:06:22Z</cim:createdDateTime>
    <cim:reason.code>A02</cim:reason.code>
    <cim:MktActivityRecord>
        <cim:mRID>87bd35cb50f946e894aa686332549d12</cim:mRID>
        <cim:originalTransactionIDReference_MktActivityRecord.mRID>OpId2022-05-16T10:06:20.219Z</cim:originalTransactionIDReference_MktActivityRecord.mRID>
        <cim:Reason>
            <cim:code>D02</cim:code>
            <cim:text>Business reason code UpdateMasterDataSettlement not allowed together with document type RequestChangeOfPriceList</cim:text>
        </cim:Reason>
    </cim:MktActivityRecord>
</cim:RejectRequestChangeOfPriceList_MarketDocument>
````

*Used D12 as document type, response:*

````
<?xml version="1.0" encoding="utf-8"?>
<cim:RejectRequestChangeOfPriceList_MarketDocument xmlns:cim="urn:ediel.org:structure:rejectrequestchangeofpricelist:0:1">
    <cim:mRID>e5d0b1410b9843e28e7746f233dd2691</cim:mRID>
    <cim:type>D11</cim:type>
    <cim:process.processType>D18</cim:process.processType>
    <cim:businessSector.type>23</cim:businessSector.type>
    <cim:sender_MarketParticipant.mRID codingScheme="A10">5790001330552</cim:sender_MarketParticipant.mRID>
    <cim:sender_MarketParticipant.marketRole.type>DDZ</cim:sender_MarketParticipant.marketRole.type>
    <cim:receiver_MarketParticipant.mRID codingScheme="A10">8100000000207</cim:receiver_MarketParticipant.mRID>
    <cim:receiver_MarketParticipant.marketRole.type>DDM</cim:receiver_MarketParticipant.marketRole.type>
    <cim:createdDateTime>2022-05-16T10:07:28Z</cim:createdDateTime>
    <cim:reason.code>A02</cim:reason.code>
    <cim:MktActivityRecord>
        <cim:mRID>bcc6bd147fcf449baa4941b4b2ace7fb</cim:mRID>
        <cim:originalTransactionIDReference_MktActivityRecord.mRID>OpId2022-05-16T10:07:27.910Z</cim:originalTransactionIDReference_MktActivityRecord.mRID>
        <cim:Reason>
            <cim:code>D02</cim:code>
            <cim:text>Document type NotifyPriceList not allowed together with business reason code UpdateChargeInformation</cim:text>
        </cim:Reason>
    </cim:MktActivityRecord>
</cim:RejectRequestChangeOfPriceList_MarketDocument>
````
